pkg/pushsync/pushsync.go

```diff
@@ -77,54 +81,78 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and inserts it in to the local store,
+// It also sends a receipt for the chunk on the same stream.
+// ALso it forwards the received chunk to its closest peer.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	// Receive the Delivery of a chunk
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
 	var ch pb.Delivery
-
 	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
+		ps.metrics.ReceivedChunkErrorCounter.Inc()
+		return err
+	}
+	ps.metrics.ChunksReceivedCounter.Inc()
+
+	// Select the closest peer to forward the received chunk
+	peer, err := ps.peerSuggester.ClosestPeer(swarm.NewAddress(ch.Address))
+	if err != nil {
+		if errors.Is(err, topology.ErrWantSelf) {
```

Comment:
Exiting here with nil means that the storer node is not sending back the receipt. This makes all intermediate nodes and the sender node never receive their own receipts.

Reply:
if the swarm book logic is followed (as specified above), this comment becomes invalid.

Reply:
it is fixed to send receipt now since the above comment was accepted.